### DISCUSSION TITLE
SAS on Demand Connection Difficulties When Flask App Deployed to Azure Web App

### DISCUSSION THREAD

**Discussion by sadovsd** [sassoftware/saspy#548] [Q&A] [answered]
When I test my flask API using a local host http://127.0.0.1:5001/, everything works fine. I followed the instruction here https://support.sas.com/ondemand/saspy.html for creating the sascfg_personal.py and .authinfo files. 

However, when testing the link of the deployed app, I get an internal server error. According to the azure log stream, there are problems accessing the .authinfo file. I included images of these error. I don't know what I have to do to properly configure and authenticate the sas session in this situation. In the repository that is getting deployed, I added the .authinfo file in a folder called root (a previous error in the log stream said it was searching for .authinfo in root). I added the sascfg_personal.py file inside the repository as well. 

For reference, the operating system on the azure web app is linux and the runtime stack is python 3.9.

<img width="172" alt="Screenshot 2023-07-13 at 9 37 21 AM" src="https://github.com/sassoftware/saspy/assets/117796771/2c7c366a-7136-47dd-8567-e49af5037606">

<img width="795" alt="Screenshot 2023-07-13 at 9 36 19 AM" src="https://github.com/sassoftware/saspy/assets/117796771/75b1bb5f-0ac4-4d4a-ab86-a48b640df84e">

<img width="786" alt="Screenshot 2023-07-13 at 9 37 03 AM" src="https://github.com/sassoftware/saspy/assets/117796771/978ad458-e315-413f-85dd-295952913959">

**Comment by tomweber-sas** [comment]
Hey, I can help you with this. There's more than one way to work around something like this. First, this looks cool, looks like you have it working and all. The .authinfo file has to be in the users home directory, and it need permissions of 600 (rw- --- ---); only the user can access it. In a container, you are the root user, not whatever your user might be on you pc or anywhere else. So, the file needs to be in the users home directory. In the container that's /root. So, it looks like you created a directory named root down in the middle of wherever that content you have lives. I can't see from here what that filesystem path is. So, since it's not /root/.authinfo, it won't be found. If you can put it in the right place it ought to be picked up. If you can't make it be in the right place, you could just include the credentials in the config file. Since this is a container, that no one would have access to, to be poking around to find those in that , or any file, that would just work. Or, since this is python, you could just read them into env vars, or just python vars before issuing SASsession() and provide them as variables in that call. 
The easiest, would be to make the file be in /root if you can do that when you build the container.

**Reply by sadovsd** [reply]
Hey Tom, thanks for the response. I'm new to software development so I don't understand the terminology completely. I'm deploying the app using a GitHub repository. Is this repository considered a container? What do you mean when you say you can't see the filesystem path. Are there supposed to be parent directories of /root/.authinfo?

Setting the credentials in the config file seems like a simple workaround. How can I modify my sascfg_personal.py file to accomplish this?

**Comment by tomweber-sas** [comment]
Hey, no problem. So, in your repo, is that where you have your sascfg_personal.py file and all of the content shown in the picture above; the FLASK_BACKEND2 pic? That looks like a directory structure. I'm assuming in your config file you have something like this:
```
oda      = {'java'      : '/usr/bin/java',
             'iomhost'  : ['odaws01-usw2-2.oda.sas.com','odaws02-usw2-2.oda.sas.com',
                           'odaws01-usw2.oda.sas.com','odaws02-usw2.oda.sas.com','odaws03-usw2.oda.sas.com','odaws04-usw2.oda.sas.com'],
            'iomport'   : 8591,
            'authkey'   : 'odatom',
            }
```
So, to provide credentials there, you would just comment out the authkey, and add omruser and omrpw to the config def you have. https://sassoftware.github.io/saspy/configuration.html#remote

Where are you running this? How are you running this? I assumed it's running in a container since the log you showed has it running in linux as root.

**Reply by sadovsd** [reply]
Just tried that and it seems to be moving me along. Now there's an issue with java not being found... I'm not sure how to get java in this project and how to store it. On the azure web app, I'm using a runtime stack of python 3.9.

To clarify, is the SAS_config_names=['oda'] line still supposed to be in the top of the sascfg_personal.py file?

I'm not sure about the specifics about how this flask app is running - I just made a new azure web app. But here is some log output from my new deployment that I think has relevant information:

<img width="874" alt="Screenshot 2023-07-13 at 1 50 56 PM" src="https://github.com/sassoftware/saspy/assets/117796771/65560d0f-800c-4619-8e09-820961dfc512">

<img width="780" alt="Screenshot 2023-07-13 at 1 51 12 PM" src="https://github.com/sassoftware/saspy/assets/117796771/ed799093-34f1-40b5-9099-7bc009350654">

**Comment by tomweber-sas** [comment]
Good deal! And yes, you still need sascfg_personal to have the SAS_config_names variable: https://sassoftware.github.io/saspy/configuration.html#sascfg-personal-py-details

I've never used that service you're using to run your app in, but the documentation for it is in that fist line in that log: http://aka.ms/webapp-linux

You may need to read that doc to understand how it works. You ought to be able to log into that container once it's up and explore it to see if/where java is, so you can provide the correct path to it. It may say where it's installed in the doc, don't know. 

Looks like you're close. That service is doing a ton of work for you, but then that means it's just a bunch of magic you aren't aware of, so you don't know what's really going on or how it's working. Still, the doc should help that part. It looks like you have a requirements.txt file in there, so if java isn't already in your container, which it may well be, you ought to be able to add it by just adding java to that file. You still need to know where it's installed so you can put that in your configurations definition.

**Comment by tomweber-sas** [comment]
Sure, it lets you ssh into the container, so you can see where java is installed. If it's not, add java to the requirements.txt file you have. Here's a link to their configuration doc for a python app, this section is about ssh'ing into the running container: https://learn.microsoft.com/en-us/azure/app-service/configure-language-python#open-ssh-session-in-browser

**Comment by tomweber-sas** [comment]
from inside the container, you can run these commands to see what java is there and where it is, I've just run it on one of my development machines to show you:
```
tom64-7> which java
/bin/java
tom64-7> java -version
openjdk version "1.8.0_342"
OpenJDK Runtime Environment (build 1.8.0_342-b07)
OpenJDK 64-Bit Server VM (build 25.342-b07, mixed mode)
tom64-7>
```

**Reply by sadovsd** [reply]
I got into the container and found out that there is no java inside, which is probably because I specified the azure web app to have a python runtime stack.

I downloaded docker and started learning how to make a custome container that would include both python and java. Still have a ways to go before I figure all that out.

I tried adding java to the requirements.txt file, but got a build error. I think java is getting treated as a python dependency.

Apparently, you can use JPype to run java code in python, but I suspect that approach to establishing a SAS connection will be far more complicated than building a new container.

**Comment by tomweber-sas** [comment]
Oh, of course requirements is only for python. I was thinking docker build dependencies. But, I dug around and found this, maybe this is the answer? https://learn.microsoft.com/en-us/answers/questions/991247/need-to-have-java-while-running-python-webapp-on-l

**Reply by sadovsd** [reply]
I tried that out. Don't have any more 500 internal server errors when testing with postman. However, based on the output, it's not doing the required SAS operations and for some reason the azure web app log stream isn't showing logs for the app itself anymore.

Will try the docker approach tomorrow.

**Comment by sadovsd** [comment, marked as answer]
I made a docker container to house both python and java and it all worked out when I deployed to azure web apps. The other change I made was not using the .authinfo file for authentication, and just entering credentials into the sascfg_personal.py file, as Tom suggested in an answer above. This is the docker file that worked for me.

```
# Get Linux
FROM alpine:3.14

# Get Java via the package manager
RUN apk update \
&& apk upgrade \
&& apk add --no-cache bash \
&& apk add --no-cache --virtual=build-dependencies unzip \
&& apk add --no-cache curl \
&& apk add --no-cache openjdk8-jre

# Get Python, PIP
RUN apk add --no-cache python3 \
&& python3 -m ensurepip \
&& pip3 install --upgrade pip setuptools \
&& rm -r /usr/lib/python*/ensurepip && \
if [ ! -e /usr/bin/pip ]; then ln -s pip3 /usr/bin/pip ; fi && \
if [[ ! -e /usr/bin/python ]]; then ln -sf /usr/bin/python3 /usr/bin/python; fi && \
rm -r /root/.cache

# Copy application code
COPY . /app

# set up working directory
WORKDIR /app

# Install community version of pandas, since listing it in requirement.txt takes forever to build
RUN echo "@community http://dl-cdn.alpinelinux.org/alpine/v3.14/community" >> /etc/apk/repositories && \
    apk --no-cache add py3-pandas@community

# Install Python dependencies
RUN pip install -r requirements.txt

# Set the entry point for running the appliation - for deploying to cloud
CMD ["gunicorn", "-b", "0.0.0.0:8080", "app:app"]
```

**Comment by tomweber-sas** [comment]
That's awesome, way to go! You said:
` I'm new to software development`
but you just did a bunch of different things that are part of modern day software development and worked through figuring it out, way to go! What you ought to do next is use the .authinfo file. Since you're building your own container, you can write that to the /root home directory, where you couldn't before, but can in your container, so it will be used there.
